core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala

```
@@ -93,6 +154,31 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
     } {
       out.close()
     }
+
+    val dataFile = getDataFile(shuffleId, mapId)
+    // Note: there is only one IndexShuffleBlockResolver per executor
+    synchronized {
+      val existedLengths = checkIndexAndDataFile(indexFile, dataFile, lengths.length)
+      if (existedLengths != null) {
+        // Use the lengths of existed output for MapStatus
+        System.arraycopy(existedLengths, 0, lengths, 0, lengths.length)
+        dataTmp.delete()
+        indexTmp.delete()
+      } else {
+        if (indexFile.exists()) {
+          indexFile.delete()
+        }
+        if (dataFile.exists()) {
+          dataFile.delete()
+        }
+        if (!indexTmp.renameTo(indexFile)) {
+          throw new IOException("fail to rename data file " + indexTmp + " to " + indexFile)
+        }
+        if (dataTmp != null && dataTmp.exists() && !dataTmp.renameTo(dataFile)) {
+          throw new IOException("fail to rename data file " + dataTmp + " to " + dataFile)
+        }
```

should we check
```
checkIndexAndDataFile(indexTmp, dataTmp, lengths.length) != null
```
what happens if this fails? Should we throw an exception?